plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go

```diff
@@ -0,0 +1,532 @@
+	var lastTags map[string]string
+	var lastAliasPath string
+	for _, update := range response.Update.Update {
+		// Prepare tags from prefix
+		currentTags := make(map[string]string, len(tags))
+		for key, val := range tags {
+			currentTags[key] = val
+		}
+		currentAliasPath, currentFields := c.handleTelemetryField(update, currentTags, &builder)
+
+		// Inherent valid alias from prefix parsing
+		if len(aliasPath) > 0 && len(currentAliasPath) == 0 {
+			currentAliasPath = aliasPath
+		}
+
+		// If tags have not changed aggregate fields, if not commit existing before starting anew
+		tagsChanged := len(lastTags) != len(currentTags)
+		for key, val1 := range lastTags {
+			if val2, ok := currentTags[key]; !ok || val2 != val1 {
+				tagsChanged = true
+				break
+			}
+		}
+		if (tagsChanged || currentAliasPath != lastAliasPath) && len(fields) > 0 {
+			c.addFields(builder.String(), lastAliasPath, fields, lastTags, timestamp)
+			fields = make(map[string]interface{})
+		}
+
+		// Copy measurements to aggregated field-map
+		for key, val := range currentFields {
+			fields[key[len(currentAliasPath)+1:]] = val
+		}
+
+		lastTags = currentTags
+		lastAliasPath = currentAliasPath
+	}
+
+	// Add final measurements
+	if len(fields) > 0 {
+		c.addFields(builder.String(), lastAliasPath, fields, lastTags, timestamp)
+	}
+}
+
+// HandleTelemetryField and add it to a measurement
+func (c *CiscoTelemetryGNMI) handleTelemetryField(update *gnmi.Update, tags map[string]string, builder *bytes.Buffer) (string, map[string]interface{}) {
+	namelen := builder.Len()
+	aliasPath := c.handlePath(update.Path, tags, builder)
+
+	var value interface{}
+	var jsondata []byte
+
+	switch val := update.Val.Value.(type) {
+	case *gnmi.TypedValue_AsciiVal:
+		value = val.AsciiVal
+	case *gnmi.TypedValue_BoolVal:
+		value = val.BoolVal
+	case *gnmi.TypedValue_BytesVal:
+		value = val.BytesVal
+	case *gnmi.TypedValue_DecimalVal:
+		value = val.DecimalVal
+	case *gnmi.TypedValue_FloatVal:
+		value = val.FloatVal
+	case *gnmi.TypedValue_IntVal:
+		value = val.IntVal
+	case *gnmi.TypedValue_StringVal:
+		value = val.StringVal
+	case *gnmi.TypedValue_UintVal:
+		value = val.UintVal
+	case *gnmi.TypedValue_JsonIetfVal:
+		jsondata = val.JsonIetfVal
+	case *gnmi.TypedValue_JsonVal:
+		jsondata = val.JsonVal
+	}
+
+	name := strings.Replace(builder.String(), "-", "_", -1)
+	fields := make(map[string]interface{})
+	if value != nil {
+		fields[name] = value
+	} else if jsondata != nil {
+		if err := json.Unmarshal(jsondata, &value); err != nil {
+			c.acc.AddError(fmt.Errorf("failed to parse JSON value: %v", err))
+		} else {
+			flattener := jsonparser.JSONFlattener{Fields: fields}
+			flattener.FullFlattenJSON(name, value, true, true)
+		}
+	}
+	builder.Truncate(namelen)
+	return aliasPath, fields
+}
+
+// Parse path to path-buffer and tag-field
+func (c *CiscoTelemetryGNMI) handlePath(path *gnmi.Path, tags map[string]string, builder *bytes.Buffer) string {
```

Comment:
The life of the builder is very difficult for me to follow, as it moves into several functions and is truncated.  Instead of passing a bytes.Buffer in, I suggest creating a new bytes.Buffer in this function and returning the results of .String().  You could pass the prefix in to handle the individual updates.

The function would return the full path and the aliased path and add any path tags.

Reply:
Got rid of passing around of the buffer.